# 01_Setup



In [ ]:
# ============================================================
# Step 1. Google Drive 연결 및 프로젝트 경로 설정
# ============================================================

from google.colab import drive
from pathlib import Path
import os

drive.mount('/content/drive')

BASE_DIR = Path("/content/drive/MyDrive/datamining_project")

DATA_RAW = BASE_DIR / "data" / "raw"
DATA_PROCESSED = BASE_DIR / "data" / "processed"
RESULTS_DIR = BASE_DIR / "results"
FIGURES_DIR = RESULTS_DIR / "figures"
MODEL_DIR = BASE_DIR / "models"

for p in [DATA_RAW, DATA_PROCESSED, RESULTS_DIR, FIGURES_DIR, MODEL_DIR]:
    p.mkdir(parents=True, exist_ok=True)

print("BASE_DIR:", BASE_DIR)
print("DATA_RAW:", DATA_RAW)
print("FIGURES_DIR:", FIGURES_DIR)
print("MODEL_DIR:", MODEL_DIR)

print("\n[파일 존재 확인]")
print("train.csv:", (DATA_RAW / "train.csv").exists())
print("valid.csv:", (DATA_RAW / "valid.csv").exists())
print("test.csv :", (DATA_RAW / "test.csv").exists())

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
BASE_DIR: /content/drive/MyDrive/datamining_project
DATA_RAW: /content/drive/MyDrive/datamining_project/data/raw
FIGURES_DIR: /content/drive/MyDrive/datamining_project/results/figures
MODEL_DIR: /content/drive/MyDrive/datamining_project/models

[파일 존재 확인]
train.csv: True
valid.csv: True
test.csv : True


# 02_Data_Load

In [ ]:
# ============================================================
# Step 1. Train / Valid / Test 데이터 로드
# ============================================================

import pandas as pd
import numpy as np

train_df = pd.read_csv(DATA_RAW / "train.csv")
valid_df = pd.read_csv(DATA_RAW / "valid.csv")
test_df  = pd.read_csv(DATA_RAW / "test.csv")

# 날짜 컬럼 변환
for df in [train_df, valid_df, test_df]:
    df["date"] = pd.to_datetime(df["date"])

print("[데이터 로드 완료]")
print("train:", train_df.shape)
print("valid:", valid_df.shape)
print("test :", test_df.shape)

print("\n[컬럼 목록]")
print(train_df.columns.tolist())

display(train_df.head())

[데이터 로드 완료]
train: (44268, 9)
valid: (5580, 9)
test : (5580, 9)

[컬럼 목록]
['date', 'adj_close', 'close', 'high', 'low', 'open', 'volume', 'ticker', 'group']


,date,adj_close,close,high,low,open,volume,ticker,group
0,2023-12-12,21594.337891,23300.0,23450.0,23050.0,23150.0,202001,000080.KS,Medium
1,2023-12-13,21594.337891,23300.0,23500.0,23100.0,23350.0,152072,000080.KS,Medium
2,2023-12-14,21640.677734,23350.0,23500.0,23200.0,23450.0,241168,000080.KS,Medium
3,2023-12-15,21408.978516,23100.0,23400.0,23100.0,23350.0,183231,000080.KS,Medium
4,2023-12-18,21501.660156,23200.0,23300.0,22950.0,23100.0,128306,000080.KS,Medium


In [ ]:
# ============================================================
# Step 2 Feature_engineering 데이터 로드
# ============================================================


print("[DATA_PROCESSED 경로]")
print(DATA_PROCESSED)

print("\n[processed 폴더 내부 파일 목록]")
for f in os.listdir(DATA_PROCESSED):
    print(f)



train_df = pd.read_csv(DATA_PROCESSED / "train_features.csv")
valid_df = pd.read_csv(DATA_PROCESSED / "valid_features.csv")
test_df  = pd.read_csv(DATA_PROCESSED / "test_features.csv")

# date 컬럼이 있으면 datetime으로 변환
for df in [train_df, valid_df, test_df]:
    if "date" in df.columns:
        df["date"] = pd.to_datetime(df["date"])

print("[데이터 로드 완료]")
print("train:", train_df.shape)
print("valid:", valid_df.shape)
print("test :", test_df.shape)

print("\n[컬럼 목록]")
print(train_df.columns.tolist())

display(train_df.head())

[DATA_PROCESSED 경로]
/content/drive/MyDrive/datamining_project/data/processed

[processed 폴더 내부 파일 목록]
test_features.csv
train_features.csv
valid_features.csv
[데이터 로드 완료]
train: (42400, 17)
valid: (3807, 17)
test : (3701, 17)

[컬럼 목록]
['date', 'adj_close', 'close', 'high', 'low', 'open', 'volume', 'ticker', 'group', 'vol_chg_rate', 'volume_ma20_ratio', 'daily_return', 'volatility_5d', 'drawdown_after_peak_5d', 'upper_shadow_ratio', 'body_ratio', 'upper_shadow_streak_5d']


,date,adj_close,close,high,low,open,volume,ticker,group,vol_chg_rate,volume_ma20_ratio,daily_return,volatility_5d,drawdown_after_peak_5d,upper_shadow_ratio,body_ratio,upper_shadow_streak_5d
0,2024-01-11,20674.576172,21400.0,22150.0,21200.0,22050.0,330021.0,000080.KS,Medium,1.539190,1.827646,-0.029478,0.012567,0.050998,0.105263,0.684211,2.0
1,2024-01-12,21061.017578,21800.0,21850.0,21400.0,21400.0,194748.0,000080.KS,Medium,-0.409892,1.080679,0.018692,0.017191,0.026786,0.111111,0.888889,1.0
2,2024-01-15,21350.847656,22100.0,22150.0,21750.0,21900.0,24583.0,000080.KS,Medium,-0.873770,0.141416,0.013761,0.018892,0.011186,0.125000,0.500000,0.0
3,2024-01-16,21012.712891,21750.0,22100.0,21700.0,22050.0,64781.0,000080.KS,Medium,1.635195,0.392576,-0.015837,0.020116,0.020270,0.125000,0.750000,0.0
4,2024-01-17,20626.271484,21350.0,21800.0,21300.0,21800.0,93005.0,000080.KS,Medium,0.435683,0.579456,-0.018391,0.021222,0.036117,0.000000,0.900000,0.0


# 03_Preprocessing

In [ ]:
# ============================================================
# Step 1. 전처리 상태 확인
# ============================================================

print("[Train]")
print("inf 개수 :", np.isinf(train_df.select_dtypes(include=np.number)).sum().sum())
print("NaN 개수 :", train_df.isna().sum().sum())
print("거래량 0 개수 :", (train_df["volume"] == 0).sum())

print("\n[Valid]")
print("inf 개수 :", np.isinf(valid_df.select_dtypes(include=np.number)).sum().sum())
print("NaN 개수 :", valid_df.isna().sum().sum())
print("거래량 0 개수 :", (valid_df["volume"] == 0).sum())

print("\n[Test]")
print("inf 개수 :", np.isinf(test_df.select_dtypes(include=np.number)).sum().sum())
print("NaN 개수 :", test_df.isna().sum().sum())
print("거래량 0 개수 :", (test_df["volume"] == 0).sum())

[Train]
inf 개수 : 0
NaN 개수 : 8
거래량 0 개수 : 0

[Valid]
inf 개수 : 0
NaN 개수 : 4
거래량 0 개수 : 0

[Test]
inf 개수 : 0
NaN 개수 : 2
거래량 0 개수 : 0


In [ ]:
# ============================================================
# Step 2. 전체 모델 공통 결측치 제거
# ============================================================

preprocessing_features = [
    "volume_ma20_ratio",
    "vol_chg_rate",
    "daily_return",
    "volatility_5d",
    "drawdown_after_peak_5d",
    "upper_shadow_ratio",
    "body_ratio",
    "upper_shadow_streak_5d"
]

train_df = (
    train_df
    .dropna(subset=preprocessing_features)
    .reset_index(drop=True)
)

valid_df = (
    valid_df
    .dropna(subset=preprocessing_features)
    .reset_index(drop=True)
)

test_df = (
    test_df
    .dropna(subset=preprocessing_features)
    .reset_index(drop=True)
)

print("===== 결측치 제거 후 =====")
print("train :", train_df.shape)
print("valid :", valid_df.shape)
print("test  :", test_df.shape)

print("\n===== 남은 NaN 개수 =====")
print("train :", train_df[preprocessing_features].isna().sum().sum())
print("valid :", valid_df[preprocessing_features].isna().sum().sum())
print("test  :", test_df[preprocessing_features].isna().sum().sum())

===== 결측치 제거 후 =====
train : (42396, 17)
valid : (3805, 17)
test  : (3700, 17)

===== 남은 NaN 개수 =====
train : 0
valid : 0
test  : 0


# 05_Baseline2_KMeans_Only

## baseline2 모델 확정

In [ ]:
# ============================================================
# 05_Baseline2_KMeans_Only
# Step 1. Baseline2 데이터 복사 및 Feature 정의
# ============================================================

base2_train = train_df.copy()
base2_valid = valid_df.copy()
base2_test  = test_df.copy()

# Baseline2는 이상치 탐지 없이 전체 데이터를 바로 군집화한다.
# 군집화 feature는 Baseline1, Main Model과 동일하게 common_cluster_features를 사용한다.
base2_cluster_features = common_cluster_features

print("Baseline2 Cluster Features")
print(base2_cluster_features)

print("\n[데이터 크기]")
print("base2_train:", base2_train.shape)
print("base2_valid:", base2_valid.shape)
print("base2_test :", base2_test.shape)

Baseline2 Cluster Features
['volume_ma20_ratio', 'vol_chg_rate', 'daily_return', 'volatility_5d', 'drawdown_after_peak_5d', 'upper_shadow_ratio', 'body_ratio', 'upper_shadow_streak_5d']

[데이터 크기]
base2_train: (42396, 17)
base2_valid: (3805, 17)
base2_test : (3700, 17)


In [ ]:
# ============================================================
# 05_Baseline2_KMeans_Only
# Step 2. Train 기준 Scaling
# ============================================================

from sklearn.preprocessing import StandardScaler

base2_scaler = StandardScaler()

X_base2_train = base2_train[base2_cluster_features]
X_base2_valid = base2_valid[base2_cluster_features]
X_base2_test  = base2_test[base2_cluster_features]

X_base2_train_scaled = base2_scaler.fit_transform(X_base2_train)
X_base2_valid_scaled = base2_scaler.transform(X_base2_valid)
X_base2_test_scaled  = base2_scaler.transform(X_base2_test)

print("Scaling 완료")
print("Train:", X_base2_train_scaled.shape)
print("Valid:", X_base2_valid_scaled.shape)
print("Test :", X_base2_test_scaled.shape)

Scaling 완료
Train: (42396, 8)
Valid: (3805, 8)
Test : (3700, 8)


In [ ]:
# ============================================================
# 05_Baseline2_KMeans_Only
# Step 3. K 후보별 Silhouette Score 비교
# ============================================================

from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

base2_k_candidates = range(2, 9)
base2_tuning_results = []

for k in base2_k_candidates:

    base2_kmeans_tmp = KMeans(
        n_clusters=k,
        random_state=42,
        n_init=10
    )

    base2_train_labels_tmp = base2_kmeans_tmp.fit_predict(
        X_base2_train_scaled
    )

    base2_silhouette_tmp = silhouette_score(
        X_base2_train_scaled,
        base2_train_labels_tmp
    )

    base2_tuning_results.append({
        "k": k,
        "silhouette": base2_silhouette_tmp,
        "inertia": base2_kmeans_tmp.inertia_
    })

base2_tuning_df = pd.DataFrame(base2_tuning_results)

display(
    base2_tuning_df.sort_values(
        by="silhouette",
        ascending=False
    )
)

,k,silhouette,inertia
0,2,0.850387,284016.662278
1,3,0.571472,250340.550054
6,8,0.263183,134265.903412
5,7,0.253582,148152.240795
4,6,0.241743,168416.392051
3,5,0.232710,194214.221875
2,4,0.230639,220422.087216


In [ ]:
# ============================================================
# Baseline2
# K = 2
# ============================================================
base2_kmeans_2 = KMeans(
    n_clusters=2,
    random_state=42,
    n_init=10
)

labels_2 = base2_kmeans_2.fit_predict(
    X_base2_train_scaled
)

pd.Series(labels_2).value_counts()

,count
0,42210
1,186


In [ ]:
pd.Series(labels_2).value_counts()

,count
0,42210
1,186


In [ ]:
base2_train["cluster"] = labels_2

display(
    base2_train
    .groupby("cluster")[base2_cluster_features]
    .mean()
)

,volume_ma20_ratio,vol_chg_rate,daily_return,volatility_5d,drawdown_after_peak_5d,upper_shadow_ratio,body_ratio,upper_shadow_streak_5d
cluster,,,,,,,,
0,1.033582,0.423000,0.000739,0.023024,0.044831,0.275331,0.545273,1.452215
1,0.730833,0.128055,0.000000,0.000000,-0.335773,-13.313285,14.147503,0.026882


In [ ]:
# ============================================================
# Baseline2 Final Candidate
# K = 3
# ============================================================

from sklearn.cluster import KMeans

base2_final_k = 3

base2_kmeans = KMeans(
    n_clusters=base2_final_k,
    random_state=42,
    n_init=10
)

base2_train["cluster"] = (
    base2_kmeans.fit_predict(
        X_base2_train_scaled
    )
)

print(
    base2_train["cluster"]
    .value_counts()
    .sort_index()
)

cluster
0     2311
1    39899
2      186
Name: count, dtype: int64


In [ ]:
# ============================================================
# Cluster Profile
# ============================================================

base2_train_profile = (
    base2_train
    .groupby("cluster")[
        base2_cluster_features
    ]
    .mean()
    .round(3)
)

display(base2_train_profile)

,volume_ma20_ratio,vol_chg_rate,daily_return,volatility_5d,drawdown_after_peak_5d,upper_shadow_ratio,body_ratio,upper_shadow_streak_5d
cluster,,,,,,,,
0,3.355,5.539,0.050,0.064,0.068,0.265,0.582,1.333
1,0.899,0.127,-0.002,0.021,0.043,0.276,0.543,1.459
2,0.731,0.128,0.000,0.000,-0.336,-13.313,14.148,0.027


In [ ]:
from sklearn.metrics import silhouette_score

base2_train_silhouette = silhouette_score(
    X_base2_train_scaled,
    base2_train["cluster"],
    sample_size=5000,
    random_state=42
)

print(
    "Train Silhouette:",
    round(base2_train_silhouette, 4)
)

Train Silhouette: 0.5839


초기 k 탐색 결과
k=2 -> silhouette = 0.85
k=3 -> silhouette = 0.57
정량적 성능만 보면 k=2 일때 가장 높다

하지만 군집 분포를 확인한 결과

k=2                count

Cluster 0   42,210
Cluster 1    186

전체 데이터의 99% 이상이 하나의 군집에 집중되었다
k=2 에서는 정상 데이터 vs 극단적인 이상 데이터 수준의 단순 분리만 이루어졌다
거래 패턴을 세부적으로 해석할 수 있는 군집 구조는 형성되지 않았다.

k=3               count

Cluster 0     2,311

Cluster 1     39,899

Cluster 2     186

일반거래형
거래량 증가형
극단 특이 패턴형
의 세가지 거래 유형으로 구분할 수 있었다.

따라서 Baseline2의 최종 k값은 3으로 결정하였다.

## Validation Dataset으로 검증


In [ ]:
# ============================================================
# Baseline2 Validation Prediction
# ============================================================

base2_valid["cluster"] = (
    base2_kmeans.predict(
        X_base2_valid_scaled
    )
)

print(
    base2_valid["cluster"]
    .value_counts()
    .sort_index()
)

cluster
0     420
1    3380
2       5
Name: count, dtype: int64


In [ ]:
from sklearn.metrics import silhouette_score

base2_valid_silhouette = silhouette_score(
    X_base2_valid_scaled,
    base2_valid["cluster"],
    sample_size=min(5000, len(base2_valid)),
    random_state=42
)

print(
    "Validation Silhouette:",
    round(base2_valid_silhouette, 4)
)

Validation Silhouette: 0.4838


In [ ]:
base2_valid_profile = (
    base2_valid
    .groupby("cluster")[
        base2_cluster_features
    ]
    .mean()
    .round(3)
)

display(base2_valid_profile)

,volume_ma20_ratio,vol_chg_rate,daily_return,volatility_5d,drawdown_after_peak_5d,upper_shadow_ratio,body_ratio,upper_shadow_streak_5d
cluster,,,,,,,,
0,2.770,2.389,0.058,0.070,0.065,0.222,0.604,1.174
1,1.030,0.082,-0.002,0.025,0.046,0.243,0.476,1.312
2,0.714,-0.250,0.000,0.000,-0.337,-6.919,7.484,0.000


## Test Dataset으로 최종 검증

In [ ]:
# ============================================================
# Baseline2 Test Prediction
# ============================================================

base2_test["cluster"] = (
    base2_kmeans.predict(
        X_base2_test_scaled
    )
)

print(
    base2_test["cluster"]
    .value_counts()
    .sort_index()
)

cluster
0     481
1    3218
2       1
Name: count, dtype: int64


In [ ]:
base2_test_silhouette = silhouette_score(
    X_base2_test_scaled,
    base2_test["cluster"],
    sample_size=min(5000, len(base2_test)),
    random_state=42
)

print(
    "Test Silhouette:",
    round(base2_test_silhouette, 4)
)

Test Silhouette: 0.4569


In [ ]:
base2_test_profile = (
    base2_test
    .groupby("cluster")[
        base2_cluster_features
    ]
    .mean()
    .round(3)
)

display(base2_test_profile)

,volume_ma20_ratio,vol_chg_rate,daily_return,volatility_5d,drawdown_after_peak_5d,upper_shadow_ratio,body_ratio,upper_shadow_streak_5d
cluster,,,,,,,,
0,2.065,1.601,0.062,0.082,0.080,0.262,0.529,1.233
1,0.931,0.069,-0.005,0.030,0.062,0.248,0.477,1.302
2,0.379,-0.403,0.000,0.000,-0.192,-10.370,10.656,0.000


Baseline 2 모델 해석

Cluster 1 (대다수)

Train : 39899개

Valid : 3380개

Test : 3218개

특징
volume_ma20_ratio ≈ 1
daily_return ≈ 0

-> 일반 거래형

Cluter 0

Train : 2311개

Valid : 420개

Test : 481개

특징

거래량 증가

수익률 증가

변동성 증가

-> 거래량 증가형

Cluster 2

Train : 186개

Valid : 5개

Test : 5개

특징

body_ratio 매우 크다

upper_shadow_ratio 극단값

-> 극단 특이 패턴형

Baseline2는 이상치 탐지 없이 전체 데이터를 직접 군집화했음에도
Validation 및 Test 데이터에서 유사한 군집 구조를 유지하였다

여기서 Main Model이 필요한 이유는 Baseline2 는 분류 수준이 낮음

정상, 약간 특이, 엄청 특이 수준의 분류만 이루어짐

반면 Main Model은 Isolation Forest + DBSCAN + KMeans 를 써서

실제 이상 거래만 선별한 뒤 그 안에서 패턴을 세분화한